6cfa84: Change aria-hidden rule to focus on tabbable rather than focusable (#1820)
act-rules/act-rules.github.io: _rules/aria-hidden-no-focusable-content-6cfa84.md

## Expectation

> None of the target elements are part of the [sequential focus navigation][], nor do they have [descendants](https://dom.spec.whatwg.org/#concept-tree-descendant) in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) that are part of the [sequential focus navigation][], unless, upon receiving focus, focus is moved to an element that does not fail this rule.

Reviewer: We might be able to rephrase like
```suggestion
None of the target elements or their [descendants][] in the [flat tree][] are part of the [sequential focus navigation][], unless, upon receiving focus, focus is moved to an element that does not fail this rule.
```
or maybe
```suggestion
None of the target elements has [inclusive descendants][] in the [flat tree][] that are part of the [sequential focus navigation][],  upon receiving focus, focus is moved to an element that does not fail this rule.
```
which does keep the text more "focused" on the targets.

> None of the target elements are part of the [sequential focus navigation][], nor do they have [descendants](https://dom.spec.whatwg.org/#concept-tree-descendant) in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) that are part of the [sequential focus navigation][], unless, upon receiving focus, focus is moved to an element that does not fail this rule.

Reviewer: I'm not sure about how the redirect exception is written currently.
```html
<div aria-hidden="true">
  <a href="…" onfocus='document.getElementById("next").focus()'>Redirect</a>
  <a href="…" id="next" tabindex="-1">Not tabbable?</a>
</div>
```
would pass the rule since the tabbable element redirect focus to an element that "does not fail the rule" (since it's not tabbable, hence ignored by the rule)

Even worse, the rule targets the `div` (the element with `aria-hidden`), not the links. So even if the second `a` was tabbable, it would "not fail the rule" since only the `div` can fail the rule.

What about "focus is moved to an element which is [included in the accessibility tree][]"?

Author: I agree. Will fix.

> None of the target elements are part of the [sequential focus navigation][], nor do they have [descendants](https://dom.spec.whatwg.org/#concept-tree-descendant) in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) that are part of the [sequential focus navigation][], unless, upon receiving focus, focus is moved to an element that does not fail this rule.

Reviewer: There is also a full can of worms around what "upon receiving focus" actually means 🙈 
Is it good if the redirection happens after 5 seconds? 0.5s? 0.5ms?
See also [this discussion](https://github.com/act-rules/act-rules.github.io/pull/1444#discussion_r761084251).

Since we somewhat have resolved this point in earlier definition, I'd suggest something like
```suggestion
None of the target elements are part of the [sequential focus navigation][], nor do they have [descendants](https://dom.spec.whatwg.org/#concept-tree-descendant) in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) that are part of the [sequential focus navigation][], unless, they are not [focusable][].
```
This is not perfect  since it doesn't look at where the focus is redirect. OTOH it would also accept elements that self-blur (without focusing anything else).

> None of the target elements are part of the [sequential focus navigation][], nor do they have [descendants](https://dom.spec.whatwg.org/#concept-tree-descendant) in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree) that are part of the [sequential focus navigation][], unless, upon receiving focus, focus is moved to an element that does not fail this rule.

Reviewer: To sum up all in a single change proposal:
```suggestion
None of the target elements has [inclusive descendants][] in the [flat tree][] that are [focusable][] and part of the [sequential focus navigation][].
```
Should catch almost the same thing and be unambiguous.

## Assumptions

> Interacting with the page does not result in changing the `aria-hidden` [attribute value][] of target elements. An example of such a situation would be when closing a modal dialog makes previously hidden and elements not [focusable][] and part of the [sequential focus navigation][] become [focusable][] and part of the [sequential focus navigation][].

Reviewer 2: ```suggestion
Interacting with the page does not result in changing the `aria-hidden` [attribute value][] of target elements. An example of such a situation would be when closing a modal dialog makes previously hidden elements  that were not [focusable][] or part of the [sequential focus navigation][] become [focusable][] and part of the [sequential focus navigation][].
```

Reviewer 2: The example here might need some work, it seems a little wordy right now with how `focusable` and `sequential focus navigation` are currently included in my opinion.

> Interacting with the page does not result in changing the `aria-hidden` [attribute value][] of target elements. An example of such a situation would be when closing a modal dialog makes previously hidden and elements not [focusable][] and part of the [sequential focus navigation][] become [focusable][] and part of the [sequential focus navigation][].

Reviewer 2: ```suggestion
Interacting with the page does not result in changing the `aria-hidden` [attribute value][] of target elements. An example of such a situation would be when closing a modal dialog makes previously hidden elements not [focusable][] and part of the [sequential focus navigation][] become [focusable][] and part of the [sequential focus navigation][].
```

> Interacting with the page does not result in changing the `aria-hidden` [attribute value][] of target elements. An example of such a situation would be when closing a modal dialog makes previously hidden and elements not part of the [sequential focus navigation][] become part of the [sequential focus navigation][].

Reviewer: We can probably keep this as is if using "focusable and part of sequential focus navigation" in expectation.

By adding `aria-hidden="true"` to an element, content authors ensure that assistive technologies will ignore the element. This can be used to hide parts of a web page that are [pure decoration](https://www.w3.org/TR/WCAG21/#dfn-pure-decoration), such as icon fonts - that are not meant to be read by assistive technologies.

> An element that is part of the [sequential focus navigation][] with `aria-hidden="true"` is ignored as part of the reading order, but still part of the focus order, making its state of [visible](#visible) or hidden unclear.

Reviewer: I'm actually not sure what this AS note meant in the first place 🤔 Notably what does visibility has to do with `aria-hidden`. Was it supposed to referred to [programatically hidden](https://act-rules.github.io/glossary/#programmatically-hidden) rather than visibility?

Seems to have been here [from the start of the rule](https://github.com/act-rules/act-rules.github.io/pull/175/files#diff-695b07cb3f307e2d42f30d9af89d1ba529ab49d749f451d4f8738edc0ee767d6R41), though… So not related to this PR at all.

Author: My reading of this is that it would be reachable by a screen reader user when tabbing, but they wouldn't find it when using other reading modes (item navigation). visible is probably the wrong word. "discoverable" maybe?

Reviewer 3: Lets try to elaborate on this a bit. It can certainly be clearer.

Author: Part of this was already said in the prior sentence. I reworded this to tie those two together better and elaborate slightly. Let me know if that makes more sense.

> However, if the element is also part of the [sequential focus navigation][], the assistive technology will want to announce the element where focus has landed. If focus lands on an element with `aria-hidden="true"`, the element's state of decorative vs non-decorative and [visible](#visible) vs hidden unclear.

Reviewer 4: Is decorative/non-decorative a state? 

Also, end of last sentence should be "is unclear" - a word is missing.

Reviewer 5: Also need to define 

> [sequential focus navigation][] 

at the bottom of the file

Author: @HelenBurge I see that already defined in this PR - it was missing in the other PR. I copied the definition from the other definitions - did I do something wrong there?

Author: @kengdoj "state" may be the wrong word, but I'm not sure of a better one offhand. `aria-hidden` is intended to indicate that the content is decorative. If it's focusable, it's interactive and therefore implied that it's not decorative. Would it be better to say "If an element with `aria-hidden="true"` can be tabbed to, it is unclear if that content is intended to be decorative or not"?

Author: @HelenBurge Disregard. I think you meant the other two definitions I missed. I'll get those.

Reviewer 4: How about:
An element with an aria-hidden attribute set to true that is also part of the [sequential focus navigation][] may cause confusion for users of assistive technologies because the element should be hidden and not [included in the accessibility tree][].

This is the [definition of included in the accessibility tree](https://act-rules.github.io/glossary/#included-in-the-accessibility-tree): 

> [Programmatically hidden](https://act-rules.github.io/glossary/#programmatically-hidden) elements are removed from the accessibility tree. However, some browsers will leave [focusable](https://act-rules.github.io/glossary/#focusable) elements with an aria-hidden attribute set to true in the accessibility tree. Because they are hidden, these elements are considered not included in the accessibility tree. This may cause confusion for users of assistive technologies because they may still be able to interact with these focusable elements using sequential keyboard navigation, even though the element should not be included in the accessibility tree.

You can see the discussion about the wording of this in https://github.com/act-rules/act-rules.github.io/pull/1684

Author: @kengdoj Does this work?
> An element with an `aria-hidden` attribute set to `true` that is also part of the [sequential focus navigation][] may cause confusion for users of assistive technologies because the element can be reached via [sequential focus navigation][], but it should be hidden and not [included in the accessibility tree][].

Reviewer 4: yes!
